src/modm/board/feather_m4/board.hpp

```diff
@@ -0,0 +1,171 @@
+ * Copyright (c) 2023, Thomas Rush
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#pragma once
+
+#include <modm/platform.hpp>
+#include <modm/debug/logger.hpp>
+
+#define MODM_BOARD_HAS_LOGGER
+
+using namespace modm::platform;
+
+namespace Board
+{
+/// @ingroup modm_board_feather_m4
+/// @{
+using namespace modm::literals;
+
+/// The following pin aliases are as labeled on the board, with the following
+/// exceptions:  Mosi is labeled "MO"; Miso is "MI"; and the 'D' pins
+/// (D4-D6, D9-D13) do not have the "D" on the silkscreen, just the numbers
+/// (except D4, go figure).
+
+/// ARef (analog reference) is jumpered DIRECTLY (via a board trace)
+/// to 3.3V, and therefore cannot be used for IO purposes.
+using ARef	= GpioA03;
+
+using A0	= GpioA02;
+using A1	= GpioA05;
+using A2	= GpioB08;
+using A3	= GpioB09;
+using A4	= GpioA04;
+using A5	= GpioA06;
+
+using Sck 	= GpioA17;
+using Mosi 	= GpioB23;
+using Miso 	= GpioB22;
+
+using Rx	= GpioB17;
+using Tx	= GpioB16;
+
+using D13	= GpioA23;
+using D12	= GpioA22;
+using D11	= GpioA21;
+using D10	= GpioA20;
+using D9	= GpioA19;
+using D6	= GpioA18;
+using D5	= GpioA16;
+using D4	= GpioA14;
+
+using Scl	= GpioA13;
+using Sda	= GpioA12;
+
+/// This is the red LED by the USB jack.
+using Led	= D13;
+
+/// The "Neopixel" is Adafruit's name for a WS2812 equivalent RGB LED.
+/// There is no breakout for it.
+using Neopixel	= GpioB03;
+
+/// Alternate names for Rx and Tx.
+using D0	= Rx;
+using D1	= Tx;
+
+/// SAMD51J19A running at 120MHz generated from the 48MHz DFLL
+/// divided by 48 for a 1MHz reference.
+struct SystemClock
+{
+	static constexpr uint32_t Frequency = 120_MHz;
+
+	static constexpr uint32_t Clock48MHz = 48_MHz;
+	static constexpr auto Generator48MHz = ClockGenerator::Generator1;
+
+	static constexpr uint32_t Clock32k = 32768;
+	static constexpr auto Generator32k = ClockGenerator::Generator2;
+
+	static constexpr uint32_t Clock1MHz = 1_MHz;
+	static constexpr auto Generator1MHz = ClockGenerator::Generator3;
+
+	static constexpr uint32_t Usb = Clock48MHz;
+
+	static constexpr uint32_t Sercom5 = Frequency;
+	static constexpr uint32_t SercomSlow = Clock32k;
+
+	static bool inline
+	enable()
+	{
+		// start the 32kHz crystal oscillator and connect to generator
+		GenericClockController::enableExternalCrystal32k(Xosc32StartupTime::Start_500ms);
+		GenericClockController::enableGenerator<Generator32k, ClockSource::Xosc32k>();
+
+		// set up a 1MHz reference for DPLL0
+		GenericClockController::enableGenerator<Generator1MHz, 
+							GeneratorConfiguration{ClockSource::Dfll, 48U}>();
+		GenericClockController::connect<ClockPeripheral::Fdpll0>(Generator1MHz);
```

Comment:
Couldn't we configure the DFLL48 in closed-loop mode with the 32k crystal as the reference?

Reply:
I had written an option into the board module.lb to choose between open and closed-loop on the DFLL48 (open-loop was the default), and then removed it. Should I put the option back, or just go with the closed-loop with 32k ref?

Reply:
If the board always has a 32k oscillator I'd go with the closed-loop mode. The USB spec has quite strict clock requirements (+- 0.25 %). The default open-loop mode won't meet the spec. Therefore, there is also the option to recover the clock from USB SOF packets and tune the DFLL48 to it, but out driver doesn't support it yet and it can only be enabled while connected to a USB host. The closed-loop mode with the 32k crystal should be the easiest option.

Reply:
That's fine with me. Here are some diffs for board.hpp and module.md to get your feedback before I commit anything else. In particular, note (in board.hpp) the definition of 'Clock1MHz' and its explanation; the value comes out to 999424Hz. I don't suppose it's really necessary to do it that way. The other section involves the initialization of DFLL48 in closed-loop; the CPU clock is temporarily changed to XOSC32K for the changeover. (This git editor won't display the diffs the way I want, but hopefully you get the idea.)

board.hpp:
90c90,93
<     static constexpr uint32_t Clock1MHz = 1_MHz;
---
> 	// when DFLL is in closed loop mode, the "1_MHz" DPLL reference is not
> 	// exactly 1MHz; calculating the actual value allows 'enableDpll()' to
> 	// check the tolerance
> 	static constexpr uint32_t Clock1MHz = ((48_MHz / 32768) * 32768) / 48;
> 103a107,118
> 
> 		// set up Xosc32k as DFLL reference
> 		GenericClockController::connect<ClockPeripheral::Dfll48>(Generator32k);
> 
> 		// while restarting DFLL, run system temporarily at 32kHz
> 		GenericClockController::enableGenerator<ClockGenerator::Main, ClockSource::Xosc32k>(); > 
> 
> 		// restart DFLL in closed loop
> 		GenericClockController::enableDfll48mClosedLoop<Clock32k>();
> 
> 		// get system back up to speed
> 		GenericClockController::enableGenerator<ClockGenerator::Main, ClockSource::Dfll>();

module.md:
58,59c58,60
< (e.g. 1-2MHz). This is the method used by Adafruit and TinyUSB in their
< FeatherM4 libraries, and that is what is done here.

> (e.g. 1-2MHz). It can be done with the DFLL in open- or closed-loop mode, but
> in order to meet the USB timing specs, closed-loop with XOSC32K as reference is
> used here.